pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+        InactivityTrackingEnabled { is_enabled: bool },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        MaxCollatorsPerSessionReached,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::DbWeight::get().reads_writes(0, 1))]
+        pub fn set_inactivity_tracking_status(
+            origin: OriginFor<T>,
+            is_enabled: bool,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            <EnableInactivityTracking<T>>::put(is_enabled);
+            Self::deposit_event(Event::<T>::InactivityTrackingEnabled { is_enabled });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
+                total_weight +=
+                    Self::process_ended_session(current_unprocessed_session, current_session);
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn process_ended_session(
+            unprocessed_session_id: SessionIndex,
+            active_session_id: SessionIndex,
+        ) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 3);
+            ActiveCollators::<T>::insert(
+                unprocessed_session_id,
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedVec::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            let minimum_sessions_required = T::MaxInactiveSessions::get() + 1;
+            if active_session_id >= minimum_sessions_required {
+                total_weight += T::DbWeight::get().writes(1);
+                let _ = <crate::pallet::ActiveCollators<T>>::remove(
+                    active_session_id - minimum_sessions_required,
+                );
+            }
+            <LastUnprocessedSession<T>>::put(active_session_id);
+            total_weight
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
```

Comment:
maybe can be inserted ordered so that the lookup takes less time? @ParthDesai and @tmpolaczyk

Reply:
Using BoundedBTreeSet instead of BoundedVec will automatically ensure it's ordered and speed up lookups

Reply:
Good point! Updated it in https://github.com/moondance-labs/tanssi/pull/922/commits/86f1ce1ef64412ec230fdb25c0ce1fa4e260578d